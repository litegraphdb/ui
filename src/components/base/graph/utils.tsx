import { NodeData } from '@/lib/graph/types';
import { defaultNodeColor, nodeLightColorByType, universalNodeColors } from './constant';

const maxLegends = 16;

export const getLegendsForNodes = (nodes: NodeData[]) => {
  const priorityLegends = new Map<string, { legend: string; color: string }>();
  const customLegends = new Map<string, { legend: string; color: string }>();

  // Separate legends into priority (nodeLightColorByType) and custom
  nodes.forEach((node) => {
    if (Object.keys(nodeLightColorByType).includes(node.type)) {
      priorityLegends.set(node.type, { legend: node.type, color: nodeLightColorByType[node.type] });
    } else {
      // Only add if not already present to avoid changing colors for same legend
      if (!customLegends.has(node.type)) {
        customLegends.set(node.type, {
          legend: node.type,
          color: universalNodeColors[customLegends.size % universalNodeColors.length],
        });
      }
    }
  });

  const priorityLegendsSize = priorityLegends.size;
  // If there are nodes with unknown types, add "Unknown" to priority legends
  const isMoreThanMaxLegends = customLegends.size + priorityLegendsSize > maxLegends;

  // Convert to arrays for easier manipulation
  const priorityArray = Array.from(priorityLegends.values());
  const customArray = Array.from(customLegends.values()).slice(
    0,
    maxLegends - (priorityArray.length + (isMoreThanMaxLegends ? 1 : 0))
  );

  // Combine priority legends first, then custom legends
  const legends = [
    ...priorityArray,
    ...customArray,
    ...(isMoreThanMaxLegends ? [{ legend: 'Unknown', color: defaultNodeColor }] : []),
  ];
  const mapLegends: Record<string, { legend: string; color: string }> = {};
  legends.forEach((legend) => {
    mapLegends[legend.legend] = legend;
  });
  return mapLegends;
};
