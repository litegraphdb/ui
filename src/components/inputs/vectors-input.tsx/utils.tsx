export const convertVectorsToAPIRecord = (
  vectors?: Array<{ Model: string; Dimensionality: number; Content: string; Vectors: number[] }>
) => {
  return (
    vectors?.map((vector: any) => ({
      ...vector,
      Dimensionality: Number(vector.Dimensionality),
      Vectors: vector.Vectors || [],
    })) || []
  );
};
