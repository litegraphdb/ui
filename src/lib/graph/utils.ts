import type { NodeData, Point } from './types';

export function isPointInNode(point: Point, node: NodeData, radius: number = 10): boolean {
  const dx = point.x - node.x;
  const dy = point.y - node.y;
  return dx * dx + dy * dy <= radius * radius;
}

export function getNodeAtPosition(nodes: NodeData[], point: Point): NodeData | null {
  return nodes.find((node) => isPointInNode(point, node)) || null;
}

export function transformToOptions<T extends { GUID: string; name?: string; Name?: string }>(
  data?: T[] | null,
  labelField: keyof T = 'name' // Field to use for label in options
) {
  return data
    ? data?.map((item) => ({
        value: item.GUID,
        label: (item[labelField] as string) || item.Name || item.GUID,
      }))
    : [];
}
