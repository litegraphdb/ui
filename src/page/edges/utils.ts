import { EdgeType } from '@/lib/store/edge/types';
import { NodeType } from '@/lib/store/node/types';

// Utility to find the name by GUID
export const getNodeNameByGUID = (guid: string, nodesList: NodeType[]) => {
  if (!guid) return 'N/A';
  const node = nodesList.find((n: NodeType) => n.GUID === guid);
  return node ? node.Name : guid;
};

export const transformEdgeDataForTable = (edgesList: EdgeType[], nodesList: NodeType[]) => {
  return (
    edgesList?.map((data: EdgeType) => {
      return {
        ...data,
        FromName: getNodeNameByGUID(data.From, nodesList),
        ToName: getNodeNameByGUID(data.To, nodesList),
        key: data.GUID,
      } as EdgeType;
    }) || []
  );
};
