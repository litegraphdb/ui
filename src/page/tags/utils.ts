import { NodeType } from '@/lib/store/node/types';

import { getNodeNameByGUID } from '../edges/utils';
import { EdgeType } from '@/lib/store/edge/types';
import { TagMetaData } from 'litegraphdb/dist/types/types';
import { TagType } from '@/lib/store/tag/types';

// Utility to find the name by GUID
export const getEdgeNameByGUID = (guid: string, edgesList: EdgeType[]) => {
  if (!guid) return 'N/A';
  const edge = edgesList.find((n: EdgeType) => n.GUID === guid);
  return edge ? edge.Name : guid;
};

export const transformTagsDataForTable = (
  tagsList: TagMetaData[],
  nodesList: NodeType[],
  edgesList: EdgeType[]
) => {
  return (
    tagsList?.map((data: TagMetaData) => {
      return {
        ...data,
        NodeName: getNodeNameByGUID(data.NodeGUID, nodesList),
        EdgeName: getEdgeNameByGUID(data.EdgeGUID, edgesList),
        key: data.GUID,
      } as TagType;
    }) || []
  );
};
