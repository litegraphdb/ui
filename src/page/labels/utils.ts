import { NodeType } from '@/types/types';
import { getNodeNameByGUID } from '../edges/utils';

import { getEdgeNameByGUID } from '../tags/utils';
import { LabelMetadata } from 'litegraphdb/dist/types/types';
import { EdgeType } from '@/types/types';
import { LabelMetadataForTable } from './types';

export const transformLabelsDataForTable = (
  labelsList: LabelMetadata[],
  nodesList: NodeType[],
  edgesList: EdgeType[]
) => {
  return (
    labelsList?.map((data: LabelMetadata) => {
      return {
        ...data,
        NodeName: getNodeNameByGUID(data.NodeGUID, nodesList),
        EdgeName: getEdgeNameByGUID(data.EdgeGUID, edgesList),
        key: data.GUID,
      } as LabelMetadataForTable;
    }) || []
  );
};

export const getNodeAndEdgeGUIDsByLabelList = (labelsList: LabelMetadata[]) => {
  const nodeGUIDs = labelsList.map((label) => label.NodeGUID);
  const edgeGUIDs = labelsList.map((label) => label.EdgeGUID);
  return {
    nodeGUIDs: Array.from(new Set(nodeGUIDs)),
    edgeGUIDs: Array.from(new Set(edgeGUIDs)),
  };
};
