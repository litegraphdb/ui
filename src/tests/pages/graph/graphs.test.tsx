import '@testing-library/jest-dom';
import React, { useState } from 'react';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import GraphPage from '@/app/dashboard/[tenantId]/graphs/page';
import AddEditGraph from '@/page/graphs/components/AddEditGraph';
import { renderWithRedux } from '../../store/utils';
import { createMockInitialState } from '../../store/mockStore';
import { setupServer } from 'msw/node';
import { handlers as graphHandlers } from './handler';
import { commonHandlers } from '@/tests/handler';
import { setTenant } from '@/lib/sdk/litegraph.service';
import { mockGraphData, mockTenantGUID } from '../mockData';
import LitegraphModal from '@/components/base/modal/Modal';
import LitegraphButton from '@/components/base/button/Button';
import LitegraphParagraph from '@/components/base/typograpghy/Paragraph';
import LitegraphTable from '@/components/base/table/Table';

const server = setupServer(...graphHandlers, ...commonHandlers);

describe('GraphPage', () => {
  beforeAll(() => server.listen());
  beforeEach(() => setTenant(mockTenantGUID));
  afterEach(() => {
    server.resetHandlers();
    jest.clearAllMocks();
    jest.clearAllTimers();
  });
  afterAll(() => server.close());

  const initialState = createMockInitialState();

  it('renders the graph page with title and create button', async () => {
    const { container } = renderWithRedux(<GraphPage />, initialState, true);

    await waitFor(() => {
      expect(screen.getByText(/graphs/i)).toBeVisible();
      expect(screen.getByRole('button', { name: /create graph/i })).toBeVisible();
    });

    expect(container).toMatchSnapshot('graph page with create button');
  });

  it('creates a new graph and calls API', async () => {
    const { container } = renderWithRedux(
      <AddEditGraph 
        isAddEditGraphVisible={true} 
        setIsAddEditGraphVisible={() => {}} 
        graph={null} 
      />,
      initialState,
      undefined,
      true
    );
  
    // Wait for modal to render using the test id that's now visible
    const modal = await screen.findByTestId('add-edit-graph-modal');
    expect(modal).toBeInTheDocument();
    
    expect(container).toMatchSnapshot('create graph modal open');
  
    // Wait for form fields
    const nameInput = await screen.findByTestId('graph-name-input');
    expect(nameInput).toBeInTheDocument();
  
    // Fill the form
    fireEvent.change(nameInput, { target: { value: 'New Graph' } });
  
    // Verify form value
    await waitFor(() => {
      expect(nameInput.value).toBe('New Graph');
    });
  
    // Find data input
    const dataInput = await screen.findByTestId('json-editor-textarea');
    fireEvent.change(dataInput, { target: { value: '{"graph":{}}' } });
  
    // Wait for Create button to be enabled - use role="button" to be specific
    await waitFor(() => {
      const createButton = screen.getByRole('button', { name: /create/i });
      expect(createButton).not.toBeDisabled();
    }, { timeout: 3000 });
  
    // Get the Create button specifically by role
    const createButton = screen.getByRole('button', { name: /create/i });
    expect(createButton).toBeInTheDocument();
  
    // Test form submission
    fireEvent.click(createButton);
  
    // Wait for any immediate state changes
    await waitFor(() => {
      expect(createButton).toBeInTheDocument();
    });
  
    // Verify form state
    expect(nameInput.value).toBe('New Graph');
    expect(container).toMatchSnapshot('after graph creation form submission');
  
  }, 8000);

  it('edits an existing graph', async () => {
    const initialState = createMockInitialState();
    const existingGraph = mockGraphData[0]; // Use first graph for editing
    
    // Test the AddEditGraph component directly in edit mode
    const { container } = renderWithRedux(
      <AddEditGraph 
        isAddEditGraphVisible={true} 
        setIsAddEditGraphVisible={() => {}} 
        graph={existingGraph} // Pass existing graph for editing
      />,
      initialState,
      undefined,
      true
    );
  
    // Wait for modal to render
    const modal = await screen.findByTestId('add-edit-graph-modal');
    expect(modal).toBeInTheDocument();
  
    // Wait for form to be populated with existing data
    const nameInput = await screen.findByTestId('graph-name-input');
    expect(nameInput).toBeInTheDocument();
  
    // The form should be pre-populated, but let's update it
    fireEvent.change(nameInput, { target: { value: 'Updated Graph' } });
  
    // Verify the updated value
    await waitFor(() => {
      expect(nameInput.value).toBe('Updated Graph');
    });
  
    // Find the Update button (since we're editing, it should say "Update" not "Create")
    const updateButton = await screen.findByRole('button', { name: /update/i });
    expect(updateButton).toBeInTheDocument();
  
    // Test form submission
    fireEvent.click(updateButton);
  
    // Wait for any form processing
    await waitFor(() => {
      expect(updateButton).toBeInTheDocument();
    });
  
    // Verify the form state
    expect(nameInput.value).toBe('Updated Graph');
    expect(container).toMatchSnapshot('after graph update');
  
  }, 8000);

  it('deletes a graph successfully', async () => {
    const initialState = createMockInitialState();
    const { container } = renderWithRedux(<GraphPage />, initialState, undefined, true);
  
    // Wait for page to load
    await waitFor(() => {
      expect(container).toBeInTheDocument();
    });
    
    const { container: modalContainer } = renderWithRedux(
      <LitegraphModal
        data-testid="delete-graph-modal"
        title="Are you sure you want to delete this graph?"
        centered
        open={true}
        onCancel={() => {}}
        footer={
          <LitegraphButton
            type="primary"
            onClick={() => {}}
            loading={false}
          >
            Confirm
          </LitegraphButton>
        }
      >
        <LitegraphParagraph>This action will delete graph.</LitegraphParagraph>
      </LitegraphModal>,
      initialState,
      undefined,
      true
    );
  
    // Wait for delete modal to appear
    await waitFor(() => {
      expect(screen.getByTestId('delete-graph-modal')).toBeInTheDocument();
    });
  
    expect(modalContainer).toMatchSnapshot('delete graph modal open');
  
    // Find and click confirm button
    const confirmButton = screen.getByText(/confirm/i);
    fireEvent.click(confirmButton);
  
    // Verify button was clicked
    expect(confirmButton).toBeInTheDocument();
  
    expect(modalContainer).toMatchSnapshot('after graph deletion');
  }, 8000);

  it('sorts graph list by name column', async () => {
    const initialState = createMockInitialState();
    
    // Create a test component with table that has data
    const TestGraphPageWithTable = () => {
      const [sortOrder, setSortOrder] = useState('asc');
      const [sortedData, setSortedData] = useState(mockGraphData);
  
      const handleSort = () => {
        const newOrder = sortOrder === 'asc' ? 'desc' : 'asc';
        setSortOrder(newOrder);
        
        const sorted = [...mockGraphData].sort((a, b) => {
          if (newOrder === 'asc') {
            return a.Name.localeCompare(b.Name);
          } else {
            return b.Name.localeCompare(a.Name);
          }
        });
        setSortedData(sorted);
      };
  
      return (
        <LitegraphTable
          columns={[
            {
              title: 'Name',
              dataIndex: 'Name',
              key: 'name',
              sorter: true,
              render: (name: string) => <div>{name}</div>,
            }
          ]}
          dataSource={sortedData}
          loading={false}
          rowKey="GUID"
          onChange={(pagination: any, filters: any, sorter: any) => {
            handleSort();
          }}
        />
      );
    };
  
    const { container } = renderWithRedux(<TestGraphPageWithTable />, initialState, undefined, true);
  
    // Wait for table to render
    await waitFor(() => {
      expect(screen.getByRole('table')).toBeInTheDocument();
    });
  
    // Find name column header
    const nameHeader = await screen.findByRole('columnheader', { name: /name/i });
    expect(nameHeader).toBeInTheDocument();
  
    // Click to sort
    fireEvent.click(nameHeader);
  
    // Verify table still exists after sort
    await waitFor(() => {
      expect(screen.getByRole('table')).toBeInTheDocument();
    });
  
    // Verify first row contains expected data
    const rows = screen.getAllByRole('row');
    expect(rows.length).toBeGreaterThan(1); // Header + at least one data row
  
    expect(container).toMatchSnapshot('after name column sort');
  }, 8000);

  it.skip('renders fallback on API failure', async () => {
    const { container } = renderWithRedux(<GraphPage />, initialState, true);

    await waitFor(() => {
      expect(screen.getByText(/something went wrong/i)).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot('graph page fallback');
  });
});
