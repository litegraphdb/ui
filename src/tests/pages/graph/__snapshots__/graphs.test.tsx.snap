// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphPage creates a new graph and calls API: after graph creation form submission 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`GraphPage creates a new graph and calls API: create graph modal open 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`GraphPage deletes a graph successfully: after graph deletion 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  >
    <div
      class="go4109123758"
      style="left: 0px; right: 0px; display: flex; position: absolute; transition: all 230ms cubic-bezier(.21,1.02,.73,1); transform: translateY(0px); top: 0px; justify-content: center;"
    >
      <div
        class="go2072408551"
        style="opacity: 0;"
      >
        <div
          class="go685806154"
        >
          <div
            class="go1858758034"
          />
          <div
            class="go1579819456"
          >
            <div
              class="go2344853693"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          class="go3958317564"
          role="status"
        >
          Add Graph successfully
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GraphPage deletes a graph successfully: delete graph modal open 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  >
    <div
      class="go4109123758"
      style="left: 0px; right: 0px; display: flex; position: absolute; transition: all 230ms cubic-bezier(.21,1.02,.73,1); transform: translateY(0px); top: 0px; justify-content: center;"
    >
      <div
        class="go2072408551"
        style="opacity: 0;"
      >
        <div
          class="go685806154"
        >
          <div
            class="go1858758034"
          />
          <div
            class="go1579819456"
          >
            <div
              class="go2344853693"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          class="go3958317564"
          role="status"
        >
          Add Graph successfully
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GraphPage edits an existing graph: after graph update 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  >
    <div
      class="go4109123758"
      style="left: 0px; right: 0px; display: flex; position: absolute; transition: all 230ms cubic-bezier(.21,1.02,.73,1); transform: translateY(0px); top: 0px; justify-content: center;"
    >
      <div
        class="go2072408551"
        style="opacity: 0;"
      >
        <div
          class="go685806154"
        >
          <div
            class="go1858758034"
          />
          <div
            class="go1579819456"
          >
            <div
              class="go2344853693"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          class="go3958317564"
          role="status"
        >
          Add Graph successfully
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GraphPage renders the graph page with title and create button: graph page with create button 1`] = `
<div>
  <main
    class="ant-layout-content pageContainer css-dev-only-do-not-override-98ntnt"
    id="graphs"
  >
    <div
      class="pageTitleContainer ant-flex css-dev-only-do-not-override-98ntnt ant-flex-wrap-wrap ant-flex-align-center ant-flex-justify-space-between ant-flex-gap-small"
    >
      <span
        class="ant-typography css-dev-only-do-not-override-98ntnt"
        data-testid="heading"
        style="font-weight: 600; font-size: 16px;"
      >
        <div
          class="ant-flex css-dev-only-do-not-override-98ntnt ant-flex-align-center"
          style="gap: 10px;"
        >
          <span
            class="ant-typography css-dev-only-do-not-override-98ntnt"
          >
            Graphs
          </span>
          <span
            aria-label="search"
            class="anticon anticon-search cursor-pointer"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="search"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
              />
            </svg>
          </span>
        </div>
      </span>
      <button
        class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-link"
        style="font-weight: 500;"
        type="button"
      >
        <span
          class="ant-btn-icon"
        >
          <span
            aria-label="plus-square"
            class="anticon anticon-plus-square"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="plus-square"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
              />
              <path
                d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
              />
            </svg>
          </span>
        </span>
        <span>
          Create Graph
        </span>
      </button>
    </div>
    <div
      class="pageContent"
    >
      <div
        class="mb-sm ant-flex css-dev-only-do-not-override-98ntnt ant-flex-align-center ant-flex-justify-space-between"
        style="margin-top: -10px; gap: 20px;"
      />
      <div
        class="ant-table-wrapper css-dev-only-do-not-override-98ntnt"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-98ntnt"
        >
          <div>
            <div
              aria-busy="true"
              aria-live="polite"
              class="ant-spin ant-spin-spinning css-dev-only-do-not-override-98ntnt"
            >
              <span
                class="ant-spin-dot-holder"
              >
                <span
                  class="ant-spin-dot ant-spin-dot-spin"
                >
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                </span>
              </span>
            </div>
          </div>
          <div
            class="ant-spin-container ant-spin-blur"
          >
            <div
              class="ant-table ant-table-empty css-dev-only-do-not-override-98ntnt"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <table
                    style="table-layout: auto;"
                  >
                    <colgroup />
                    <thead
                      class="ant-table-thead"
                    >
                      <tr>
                        <th
                          class="ant-table-cell"
                        />
                      </tr>
                    </thead>
                    <tbody
                      class="ant-table-tbody"
                    >
                      <tr
                        class="ant-table-placeholder"
                      >
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="css-dev-only-do-not-override-98ntnt ant-empty ant-empty-normal"
                          >
                            <div
                              class="ant-empty-image"
                            >
                              <svg
                                height="41"
                                viewBox="0 0 64 41"
                                width="64"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <title>
                                  Simple Empty
                                </title>
                                <g
                                  fill="none"
                                  fill-rule="evenodd"
                                  transform="translate(0 1)"
                                >
                                  <ellipse
                                    cx="32"
                                    cy="33"
                                    fill="#f5f5f5"
                                    rx="32"
                                    ry="7"
                                  />
                                  <g
                                    fill-rule="nonzero"
                                    stroke="#d9d9d9"
                                  >
                                    <path
                                      d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                    />
                                    <path
                                      d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                      fill="#fafafa"
                                    />
                                  </g>
                                </g>
                              </svg>
                            </div>
                            <div
                              class="ant-empty-description"
                            >
                              No data
                            </div>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`GraphPage sorts graph list by name column: after name column sort 1`] = `
<div>
  <div
    class="ant-table-wrapper css-dev-only-do-not-override-98ntnt"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-98ntnt"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table css-dev-only-do-not-override-98ntnt"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      aria-sort="ascending"
                      class="ant-table-cell ant-table-column-sort ant-table-column-has-sorters"
                      scope="col"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          Name
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up active"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="d52aeab4-4de7-4076-98dd-461d4a61ac88"
                  >
                    <td
                      class="ant-table-cell ant-table-column-sort"
                    >
                      <div>
                        testttt 2
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="e6d4294e-6f49-4d67-8260-5e44c2b077a6"
                  >
                    <td
                      class="ant-table-cell ant-table-column-sort"
                    >
                      <div>
                        Test Demo Graph
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <ul
          class="ant-pagination ant-table-pagination ant-table-pagination-right css-dev-only-do-not-override-98ntnt"
        >
          <li
            aria-disabled="true"
            class="ant-pagination-prev ant-pagination-disabled"
            title="Previous Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="left"
                class="anticon anticon-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </span>
            </button>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
            tabindex="0"
            title="1"
          >
            <a
              rel="nofollow"
            >
              1
            </a>
          </li>
          <li
            aria-disabled="true"
            class="ant-pagination-next ant-pagination-disabled"
            title="Next Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="right"
                class="anticon anticon-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  >
    <div
      class="go4109123758"
      style="left: 0px; right: 0px; display: flex; position: absolute; transition: all 230ms cubic-bezier(.21,1.02,.73,1); transform: translateY(0px); top: 0px; justify-content: center;"
    >
      <div
        class="go2072408551"
        style="opacity: 0;"
      >
        <div
          class="go685806154"
        >
          <div
            class="go1858758034"
          />
          <div
            class="go1579819456"
          >
            <div
              class="go2344853693"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          class="go3958317564"
          role="status"
        >
          Add Graph successfully
        </div>
      </div>
    </div>
  </div>
</div>
`;
